### PR TITLE
PR: Add encoding handling to stderr files in the IPython Console

### LINKED ISSUE #4191: Kernel dies and can't report errors because of failure to read stderr file
## Description of your problem
Hi, I am a very inexperienced user, I launched spyder to learn write one of my first, very simple programs, but i repeatedly, once each several seconds, get message:

WARNING:traitlets:kernel died: 3.0 
Traceback (most recent call last):
  File "C:\Users\Wojtek\Anaconda2\lib\site-packages\spyder\widgets\ipythonconsole\client.py", line 357, in kernel_restarted_message
    stderr = codecs.open(self.stderr_file, 'r', encoding='utf-8').read()
  File "C:\Users\Wojtek\Anaconda2\lib\codecs.py", line 686, in read
    return self.reader.read(size)
  File "C:\Users\Wojtek\Anaconda2\lib\codecs.py", line 492, in read
    newchars, decodedbytes = self.decode(data, self.errors)
UnicodeDecodeError: 'utf8' codec can't decode byte 0xbf in position 2136: invalid start byte

**What steps will reproduce the problem?**

1. It happens just after opening spyder
2. It keeps turning to internal console to show this message, so i'm in fact unable to use spyder
3. 

**What is the expected output? What do you see instead?**

**Please provide any additional information below**

## Versions and main components

* Spyder Version: 3.1.3
* Python Version: 2.7
* Qt Version: 5.6.2
* PyQt Version: 5.6.0
* Operating system: windows 7

## Dependencies

Please go to the menu entry `Help > Optional Dependencies` (or
`Help > Dependencies`), press the button `Copy to clipboard`
and paste the contents below:

jedi =0.9.0      :  0.9.0 (OK)
matplotlib >=1.0 :  2.0.0 (OK)
nbconvert >=4.0  :  4.2.0 (OK)
numpy >=1.7      :  1.11.3 (OK)
pandas >=0.13.1  :  0.19.2 (OK)
pep8 >=0.6       :  1.7.0 (OK)
pyflakes >=0.5.0 :  1.5.0 (OK)
pygments >=2.0   :  2.1.3 (OK)
pylint >=0.25    :  1.6.4 (OK)
qtconsole >=4.2.0:  4.2.1 (OK)
rope >=0.9.4     :  0.9.4 (OK)
sphinx >=0.6.6   :  1.5.1 (OK)
sympy >=0.7.3    :  1.0 (OK)

### PR DESCRIPTION
**Pull request by dalthviz**
Fixes #4191

### CODE REVIEW

File: spyder/widgets/ipythonconsole/client.py [hunk 1]
```diff
@@ -32,6 +32,7 @@
 from spyder.config.gui import get_font, get_shortcut
 from spyder.utils import icon_manager as ima
 from spyder.utils import sourcecode
+from spyder.utils.encoding import getfilesystemencoding
```
Comment (ccordoba12) [inline spyder/widgets/ipythonconsole/client.py]: Let's use `get_coding` from the same module instead

File: spyder/widgets/ipythonconsole/client.py [hunk 2]
```diff
@@ -378,7 +386,24 @@ def restart_kernel(self):
     @Slot(str)
     def kernel_restarted_message(self, msg):
         """Show kernel restarted/died messages."""
-        stderr = codecs.open(self.stderr_file, 'r', encoding='utf-8').read()
+        try:
+            stderr = codecs.open(self.stderr_file, 'r',
+                                 encoding='utf-8').read()
+        except(UnicodeDecodeError):
```
Comment (ccordoba12) [inline spyder/widgets/ipythonconsole/client.py]: There's no need to use parentheses here.

File: spyder/widgets/ipythonconsole/client.py (hunk 1 above)
Comment (ccordoba12) [inline spyder/widgets/ipythonconsole/client.py]: Sorry, this is not correct. Let me think better about it.

File: spyder/widgets/ipythonconsole/client.py [hunk 3]
```diff
@@ -378,7 +386,25 @@ def restart_kernel(self):
...
+                                 encoding='utf-8').read()
+        except UnicodeDecodeError:
+            # This handling is needed since the stderr file could be encoded
+            # in something different to utf-8. In case of fail at least
+            # we raise an informative exception.
+            # See issue 4191
+            try:
+                encoding = get_coding(open(self.stderr_file, 'rb').read())
+                stderr = codecs.open(self.stderr_file, 'r',
+                                     encoding=encoding).read()
```
Comment (ccordoba12) [inline spyder/widgets/ipythonconsole/client.py]: Let's change this to

```python
stderr_text = open(self.stderr_file, 'rb').read()
encoding = get_coding(stderr_text)
stderr = to_text_string(stderr_text, encoding)
```

File: spyder/widgets/ipythonconsole/client.py [hunk 4]
```diff
@@ -378,7 +386,25 @@ def restart_kernel(self):
...
+            except UnicodeDecodeError:
+                try:
+                    stderr = codecs.open(self.stderr_file, 'r',
+                                         encoding='cp437').read()
+                except UnicodeDecodeError:
+                    raise EncodingError("Could not read the stderr "
+                                        "file of the kernel while restarting.")
```
Comment (ccordoba12) [inline spyder/widgets/ipythonconsole/client.py]: and remove this whole block, i.e

```python
        except UnicodeDecodeError:
                try:
                    stderr = codecs.open(self.stderr_file, 'r',
                                         encoding='cp437').read()
                except UnicodeDecodeError:
                    raise EncodingError("Could not read the stderr "
                                        "file of the kernel while restarting.")
```

and change it to

```python
except:
    stderr = None
```

The thing is `cp437` is another particular encoding, and we should have detected it in the previous block. If we're not able to read `self.stderr_file`, then we need to skip this,

**Comment by goanpeca** [comment]
Tests :-p

File: spyder/widgets/ipythonconsole/client.py [hunk 5]
```diff
@@ -69,6 +71,13 @@ def background(f):
 
 
 #-----------------------------------------------------------------------------
+# Encoding Exception
+#-----------------------------------------------------------------------------
+class EncodingError(Exception):
+    """Exception when stderr file can't be opened."""
```
Comment (ccordoba12) [inline spyder/widgets/ipythonconsole/client.py]: This is unnecessary now, please remove it.

File: spyder/widgets/ipythonconsole/client.py [hunk 6]
```diff
@@ -378,7 +387,20 @@ def restart_kernel(self):
...
+        except UnicodeDecodeError:
+            # This handling is needed since the stderr file could be encoded
+            # in something different to utf-8. In case of fail at least
+            # we raise an informative exception.
+            # See issue 4191
+            try:
+                stderr_text = open(self.stderr_file, 'rb').read()
+                encoding = get_coding(stderr_text)
+                stderr = to_text_string(stderr_text, encoding)
+            except UnicodeDecodeError:
```
Comment (ccordoba12) [inline spyder/widgets/ipythonconsole/client.py]: Please change this to only `except` here. We want to catch all errors at this point.

File: spyder/widgets/ipythonconsole/client.py [hunk 7]
```diff
@@ -378,7 +387,20 @@ def restart_kernel(self):
...
+                                 encoding='utf-8').read()
+        except UnicodeDecodeError:
+            # This handling is needed since the stderr file could be encoded
+            # in something different to utf-8. In case of fail at least
+            # we raise an informative exception.
+            # See issue 4191
+            try:
+                stderr_text = open(self.stderr_file, 'rb').read()
+                encoding = get_coding(stderr_text)
+                stderr = to_text_string(stderr_text, encoding)
```
Comment (ccordoba12) [inline spyder/widgets/ipythonconsole/client.py]: Let's move these three lines to a new method called `_read_stderr`, so we can test it independently of this one.

File: spyder/widgets/ipythonconsole/client.py [hunk 8]
```diff
@@ -378,7 +387,20 @@ def restart_kernel(self):
     @Slot(str)
     def kernel_restarted_message(self, msg):
         """Show kernel restarted/died messages."""
-        stderr = codecs.open(self.stderr_file, 'r', encoding='utf-8').read()
+        try:
+            stderr = codecs.open(self.stderr_file, 'r',
+                                 encoding='utf-8').read()
+        except UnicodeDecodeError:
+            # This handling is needed since the stderr file could be encoded
```
Comment (ccordoba12) [inline spyder/widgets/ipythonconsole/client.py]: Remove `handling` here.

File: spyder/widgets/ipythonconsole/client.py [hunk 9]
```diff
@@ -378,7 +387,20 @@ def restart_kernel(self):
...
     def kernel_restarted_message(self, msg):
         """Show kernel restarted/died messages."""
-        stderr = codecs.open(self.stderr_file, 'r', encoding='utf-8').read()
+        try:
+            stderr = codecs.open(self.stderr_file, 'r',
+                                 encoding='utf-8').read()
+        except UnicodeDecodeError:
+            # This handling is needed since the stderr file could be encoded
+            # in something different to utf-8. In case of fail at least
+            # we raise an informative exception.
```
Comment (ccordoba12) [inline spyder/widgets/ipythonconsole/client.py]: Remove `In case of fail ... an informative exception.`

**Comment by ccordoba12** [comment]
Yes, that's what's missing. The test needs to go in `plugins/tests/test_ipythonconsole.py` and it needs to go like this

1. Start a console
2. Write some message in the `cp437` encoding to its `stderr_file`.
3. Use `_read_stderr` to get those contents.
4. Assert that the result of `_read_stderr` is the same content written to `stderr_file`.

**Comment by ccordoba12** [comment]
Great work @dalthviz!! I think this is ready, so nothing else to do for me than merge ;-)

**Pull request merged**